ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1765,79 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+ULONG
+IopValidateID(IN PWCHAR Id,
+              IN SIZE_T MaxIdLen,
+              IN ULONG MaxSeparators,
+              IN BOOLEAN IsMultiSz)
+{
+   PWCHAR PtrSymbol;
+   PWCHAR StringEnd;
+   WCHAR Symbol;
+   ULONG SeparatorsCount = 0;
+   PWCHAR PtrPrevSymbol = NULL;
+
+   PAGED_CODE();
+
+   StringEnd = Id + MaxIdLen;
+
+   for (PtrSymbol = Id; PtrSymbol < StringEnd; PtrSymbol++)
+   {
+      Symbol = *PtrSymbol;
+
+      if (Symbol == UNICODE_NULL)
+      {
+         if (!IsMultiSz || (PtrPrevSymbol && PtrSymbol == PtrPrevSymbol + 1))
+         {
+            if (MaxSeparators == SeparatorsCount ||
+                MaxSeparators == MAX_SEPARATORS_MULTISZ)
+            {
+               return (ULONG)(PtrSymbol - Id) + 1;
+            }
+
+            DPRINT("IopValidateID: SeparatorsCount - %lu, MaxSeparators - %lu\n",
+                   SeparatorsCount, MaxSeparators);
+
+            // FIXME logging
+            return 0;
+         }
+
+         StringEnd += MaxIdLen;
+         PtrPrevSymbol = PtrSymbol;
+      }
+      else if (Symbol < ' ' || Symbol > 0x7F || Symbol == ',')
```

Comment:
interestingly, the MSDN link you mentioned, mentions that " Less than or equal to 0x20 (' ') "  is " illegal in an ID for this IRP".

Reply:
```
      }
      else if (Symbol == ' ')
      {
         *PtrSymbol = '_';
      }
      else if (Symbol == '\\')
```